fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/ParameterVisitorBatchDAO.java

```diff
@@ -809,66 +751,31 @@ public void visit(ReferenceParmVal rpv) throws FHIRPersistenceException {
-            //  - absolute URI ==> urn:uuid:53fefa32-1111-2222-3333-55ee120877b7
-            // stored as a token with the default system
-            rec = new ResourceTokenValueRec(parameterNameId, resourceType, resourceTypeId, logicalResourceId, TokenParmVal.DEFAULT_TOKEN_SYSTEM, valueString);
-        } else if (valueString.startsWith("#")) {
-            //  - Internal ==> #fragmentid1
-            // stored as a token value with the default system
-            rec = new ResourceTokenValueRec(parameterNameId, resourceType, resourceTypeId, logicalResourceId, TokenParmVal.DEFAULT_TOKEN_SYSTEM, valueString);
-        } else {
-            //  - Relative ==> Patient/1234
-            //  - Relative ==> Patient/1234/_history/2
-            String[] tokens = valueString.split("/");
-            if (tokens.length > 1) {
-                String refResourceType = tokens[0];
-                String refLogicalId = tokens[1];
-                Integer refVersion = null;
-                if (tokens.length == 4 && HISTORY.equals(tokens[2])) {
-                    // versioned reference
-                    refVersion = Integer.parseInt(tokens[3]);
-                }
+        // The ReferenceValue has already been processed to convert the reference to
+        // the required standard form, ready for insertion as a token value.
+        ReferenceValue refValue = rpv.getRefValue();
 
-                // Store a token value configured as a reference to another resource
-                rec = new ResourceTokenValueRec(parameterNameId, resourceType, resourceTypeId, logicalResourceId, refResourceType, refLogicalId, refVersion);
+        // Ignore references containing only a "display" element (apparently supported by the spec,
+        // but contains nothing useful to store because there's no searchable value).
+        String refResourceType = refValue.getTargetResourceType();
+        String refLogicalId = refValue.getValue();
+        Integer refVersion = refValue.getVersion();
+        ResourceTokenValueRec rec;
 
-            } else {
-                // SearchReferenceTest system integration tests require support for arbitrary reference strings
-                //  - Relative ==> 1234
-                final String codeSystem = TokenParmVal.DEFAULT_TOKEN_SYSTEM;
-                rec = new ResourceTokenValueRec(parameterNameId, resourceType, resourceTypeId, logicalResourceId, codeSystem, valueString);
-            }
+        if (refValue.getType() == ReferenceType.DISPLAY_ONLY || refValue.getType() == ReferenceType.INVALID) {
+            // protect against code regression. Invalid/improper references should be
+            // filtered out already.
+            logger.warning("Invalid reference parameter type: " + resourceType + "." + rpv.getName() + " type=" + refValue.getType().name());
+            throw new IllegalArgumentException("Invalid reference parameter value. See server log for details.");
```

Comment:
I think this should be thrown as an error only.

Reply:
See previous comment.